libs/ballot-interpreter-nh-next/src/timing_marks.rs

```diff
@@ -0,0 +1,1038 @@
+        top_left_rect,
+        top_right_rect,
+        bottom_left_rect,
+        bottom_right_rect,
+        top_rects,
+        bottom_rects,
+        left_rects,
+        right_rects,
+    } = partial_timing_marks;
+
+    let rotator = Rotator180::new(*image_size);
+
+    let (top_left_corner, top_right_corner, bottom_left_corner, bottom_right_corner) = (
+        rotator.rotate_point_f32(bottom_right_corner),
+        rotator.rotate_point_f32(bottom_left_corner),
+        rotator.rotate_point_f32(top_right_corner),
+        rotator.rotate_point_f32(top_left_corner),
+    );
+
+    let (top_left_rect, top_right_rect, bottom_left_rect, bottom_right_rect) = (
+        bottom_right_rect.map(|r| rotator.rotate_rect(&r)),
+        bottom_left_rect.map(|r| rotator.rotate_rect(&r)),
+        top_right_rect.map(|r| rotator.rotate_rect(&r)),
+        top_left_rect.map(|r| rotator.rotate_rect(&r)),
+    );
+
+    let mut rotated_top_rects: Vec<Rect> =
+        top_rects.iter().map(|r| rotator.rotate_rect(r)).collect();
+    let mut rotated_bottom_rects: Vec<Rect> = bottom_rects
+        .iter()
+        .map(|r| rotator.rotate_rect(r))
+        .collect();
+    let mut rotated_left_rects: Vec<Rect> =
+        left_rects.iter().map(|r| rotator.rotate_rect(r)).collect();
+    let mut rotated_right_rects: Vec<Rect> =
+        right_rects.iter().map(|r| rotator.rotate_rect(r)).collect();
+
+    rotated_bottom_rects.sort_by_key(Rect::left);
+    rotated_top_rects.sort_by_key(Rect::left);
+    rotated_left_rects.sort_by_key(Rect::top);
+    rotated_right_rects.sort_by_key(Rect::top);
+
+    Partial {
+        geometry,
+        top_left_corner,
+        top_right_corner,
+        bottom_left_corner,
+        bottom_right_corner,
+        top_left_rect,
+        top_right_rect,
+        bottom_left_rect,
+        bottom_right_rect,
+        top_rects: rotated_bottom_rects,
+        bottom_rects: rotated_top_rects,
+        left_rects: rotated_right_rects,
+        right_rects: rotated_left_rects,
+    }
+}
+
+#[time]
+pub fn find_complete_timing_marks_from_partial_timing_marks(
+    geometry: &Geometry,
+    partial_timing_marks: &Partial,
+    debug: &ImageDebugWriter,
+) -> Option<Complete> {
+    let top_line = &partial_timing_marks.top_rects;
+    let bottom_line = &partial_timing_marks.bottom_rects;
+    let left_line = &partial_timing_marks.left_rects;
+    let right_line = &partial_timing_marks.right_rects;
+
+    let mut all_distances = vec![];
+    all_distances.append(&mut distances_between_rects(top_line));
+    all_distances.append(&mut distances_between_rects(bottom_line));
+    all_distances.append(&mut distances_between_rects(left_line));
+    all_distances.append(&mut distances_between_rects(right_line));
+    all_distances.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));
+
+    if all_distances.is_empty() {
+        return None;
+    }
+
+    let median_distance = all_distances[all_distances.len() / 2];
```

Comment:
Is it safe to assume that the vertical median distance is the same as the horizontal median distance? Or should we compute a diff median distance for each dimension? (I'm thinking both about whether the distance is the same in each dimension in the template timing mark grid but also how different skew/scanning issues could influence the different dimensions)

Reply:
They do vary due to vertical stretching, so if you were to compute the values separately you might get slightly different values. I haven't yet seen evidence that this would meaningfully improve timing mark detection, but is something to keep in mind if we do see problems.

Reply:
Sounds good